cmd/mps-control-daemon/main.go

```diff
@@ -0,0 +1,238 @@
+**/
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/urfave/cli/v2"
+	"k8s.io/klog/v2"
+
+	"github.com/NVIDIA/k8s-device-plugin/cmd/mps-control-daemon/mount"
+	"github.com/NVIDIA/k8s-device-plugin/cmd/mps-control-daemon/mps"
+	"github.com/NVIDIA/k8s-device-plugin/internal/info"
+	"github.com/NVIDIA/k8s-device-plugin/internal/logger"
+	"github.com/NVIDIA/k8s-device-plugin/internal/rm"
+
+	spec "github.com/NVIDIA/k8s-device-plugin/api/config/v1"
+)
+
+// Config represents a collection of config options for the device plugin.
+type Config struct {
+	configFile string
+
+	// flags stores the CLI flags for later processing.
+	flags []cli.Flag
+}
+
+func main() {
+	config := &Config{}
+
+	c := cli.NewApp()
+	c.Name = "NVIDIA MPS Control Daemon"
+	c.Version = info.GetVersionString()
+	c.Action = func(ctx *cli.Context) error {
+		return start(ctx, config)
+	}
+	c.Commands = []*cli.Command{
+		mount.NewCommand(),
+	}
+
+	config.flags = []cli.Flag{
+		&cli.StringFlag{
+			Name:    "nvidia-driver-root",
+			Value:   "/",
+			Usage:   "the root path for the NVIDIA driver installation (typical values are '/' or '/run/nvidia/driver')",
+			EnvVars: []string{"NVIDIA_DRIVER_ROOT"},
+		},
+		&cli.StringFlag{
+			Name:        "config-file",
+			Usage:       "the path to a config file as an alternative to command line options or environment variables",
+			Destination: &config.configFile,
+			EnvVars:     []string{"CONFIG_FILE"},
+		},
+	}
+	c.Flags = config.flags
+
+	klog.Infof("Starting %v %v", c.Name, c.Version)
+	err := c.Run(os.Args)
+	if err != nil {
+		klog.Error(err)
+		os.Exit(1)
+	}
+}
+
+// TODO: This needs to do similar validation to the plugin.
+func validateFlags(config *spec.Config) error {
+	return nil
+}
+
+// loadConfig loads the config from the spec file.
+func (cfg *Config) loadConfig(c *cli.Context) (*spec.Config, error) {
+	config, err := spec.NewConfig(c, cfg.flags)
+	if err != nil {
+		return nil, fmt.Errorf("unable to finalize config: %w", err)
+	}
+	err = validateFlags(config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to validate flags: %w", err)
+	}
+	config.Flags.GFD = nil
+
+	return config, nil
+}
+
+// TODO: This should be reused between the device plugin and this component
+func newOSWatcher(sigs ...os.Signal) chan os.Signal {
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, sigs...)
+
+	return sigChan
+}
+
+func start(c *cli.Context, cfg *Config) error {
+	klog.Info("Starting OS watcher.")
+	sigs := newOSWatcher(syscall.SIGHUP, syscall.SIGINT, syscall.SIGTERM, syscall.SIGQUIT)
+
+	var restarting bool
+	var restartTimeout <-chan time.Time
+	var daemons []*mps.Daemon
+restart:
+	// If we are restarting, stop daemons from previous run.
+	if restarting {
+		err := stopDaemons(daemons...)
+		if err != nil {
+			return fmt.Errorf("error stopping plugins from previous run: %v", err)
+		}
+	}
+
+	klog.Info("Starting Daemons.")
```

Comment:
we will need multiple control daemons per same node?

Reply:
In the current implementation, an MPS daemon sets (or will as soon as I add that logic) the active thread percentage and device memory based on the number of replicas selected for a device. That means that if all devices are shared with the same number of replicas we would only need a single daemon.

If we are running `mixed` mig mode, or end up enabling resource renaming with more flexible sharing, we would have to run a separate daemon per configuration -- where the active thread percentage is the value that cannot be controlled per device by a single daemon. Here, I would still have to investigate what is required to have them running in the same container -- or whether we'd have to end up launching multiple containers instead.

I think for the initial relase, I would like to limit the implementation to a single daemon and I'd be happy to clean up the code here to make that clearer.

cc @klueska

Reply:
yes, that will be probably better. Thanks for clarifying.